### PR TITLE
Fall back to lazy index when repo exceeds max file limit

### LINKED ISSUE #10207: Project Explorer fails to expand populated folders in large repos after ExceededMaxFileLimit
## Summary
Project Explorer can show dot-directories (for example `.agents`) as non-expandable/empty in large repositories because file tree construction fails with `ExceededMaxFileLimit`.

## Problem
In workspace root `~/code`, the `.agents` folder is visible in Warp's sidebar but appears empty or does not expand as expected, even though it contains many files and subdirectories on disk.

## Expected behavior
Project Explorer should either (a) expand `.agents` and show its contents, or (b) show a clear, user-visible message that indexing/tree generation was limited due to repository size.

## Actual behavior
The folder appears effectively empty/non-expandable in the sidebar while filesystem checks confirm it is populated. Logs show file tree generation failures:
- `Failed to build file tree for repository: ExceededMaxFileLimit`
- repo root: `~/code`

## Reproduction steps or desired workflow
1. Open Warp with project root `~/code`.
2. In Project Explorer, try to expand `.agents`.
3. Observe it appears empty or does not show expected contents.
4. Check local filesystem and logs:
   - `.agents` contains many files/subfolders.
   - `~/Library/Logs/warp.log` includes repeated `Failed to build file tree for repository: ExceededMaxFileLimit` events for this repo root.

## Artifacts
- Log evidence from `~/Library/Logs/warp.log`:
  - `Failed to build file tree for repository: ExceededMaxFileLimit`
  - repeated occurrences tied to `~/code`
- Local repository scale at time of investigation: ~153,428 tracked files.

## Warp version
Latest

## Operating system
macOS (version unknown)

### PR DESCRIPTION
**Pull request by SagarSDagdu**
## Description

Fixes #10207. When initial repo indexing exceeded `MAX_FILES_PER_REPO` (100k), the repo landed in `IndexedRepoState::Failed` and the file tree silently rendered an empty entry, hiding populated folders like `.agents` in monorepos (the reporter has a 153k-file workspace).

Implementation follows the approved spec at [`specs/gh-10207/TECH.md`](https://github.com/warpdotdev/warp/blob/master/specs/gh-10207/TECH.md) (merged from #10490). Scope agreed with @moirahuang and @alokedesai is strictly the file tree — no toast, no banner, no event plumbing. The file tree experience should not change for users in degraded mode beyond "the folder expands."

**What:** On `Err(BuildTreeError::ExceededMaxFileLimit)` from the full-depth `Entry::build_tree`, retry once with `max_depth = 1` and `remaining_file_quota = None`. The repo installs as `IndexedRepoState::Indexed` via the existing `add_repository_internal`, which emits the usual `RepositoryUpdated` event so the view refreshes through its normal code path. Per-directory lazy-load (`LAZY_LOAD_FILE_LIMIT=5000`) handles expansion from there.

**Why no quota on the retry:** direct-child files at depth=1 still consume the file quota (`entry.rs:214-220`), so reusing `MAX_FILES_PER_REPO` would re-trigger `ExceededMaxFileLimit` on a repo with >100k files *directly* under the root and reproduce the empty-tree bug. With `None`, cost is bounded by root-entry count (subdirectories return as unloaded immediately at depth=1).

**Net diff vs master:** +43 / -4 in `crates/repo_metadata/src/local_model.rs`. No event variant, no wrapper plumbing, no view changes, no skill code.

## Linked Issue

Closes #10207

- [x] The linked issue is labeled `ready-to-spec` or `ready-to-implement`.

## Screenshots / Videos

The previously linked Loom 
https://www.loom.com/share/660730d889b6466798b52210ca6f9049 
shows the earlier iteration that included a toast. The shipped behavior here is silent — the file tree looks like any other repo, just with subfolders lazy-loaded. Verified manually against the 150k-file fixture at `~/code-fixtures/warp-10207-large-repo`; log shows `Repository exceeded max file limit; indexed in degraded mode` and the tree expands normally.

The new video recording showing logs as a proof:

https://github.com/user-attachments/assets/323bf33d-1406-4074-b666-89f208a4a553

## Testing

- Manual: 150k-file fixture in `~/code-fixtures/warp-10207-large-repo`; tree expands, log line fires, no toast.
- `cargo test -p repo_metadata --lib` — 51 passed
- `cargo test -p warp --lib code::file_tree::` — 41 passed
- `cargo test -p warp --lib ai::skills::` — 61 passed
- `cargo clippy -p repo_metadata -p warp --lib --tests -- -D warnings` — clean
- `cargo fmt --check` — clean

Unit-test follow-up tracked in spec: parameterizing `MAX_FILES_PER_REPO` (currently a `const`) is needed to drive both the original empty-tree regression and the >100k-files-at-root edge case from a test. Left out of this PR to keep the change minimal; the spec's Testing section captures the design.

## Agent Mode
- [ ] Warp Agent Mode - This PR was created via Warp's AI Agent Mode

### CODE REVIEW

File: app/src/ai/skills/file_watchers/skill_watcher.rs [hunk 1]
```diff
@@ -187,6 +187,7 @@ impl SkillWatcher {
                 | RepoMetadataEvent::RepositoryRemoved { .. }
                 | RepoMetadataEvent::FileTreeUpdated { .. }
                 | RepoMetadataEvent::UpdatingRepositoryFailed { .. }
+                | RepoMetadataEvent::RepositoryIndexedWithLimit { .. }
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply app/src/ai/skills/file_watchers/skill_watcher.rs]: Good catch. Fixed in 41e0f97a: added `find_top_level_skill_directories_in_filesystem` that probes `<repo>/<provider.skills_path>` directly via the filesystem (since the depth=1 tree omits these), and a new `RepositoryIndexedWithLimit` arm in `SkillWatcher` that calls `watch_repo` and feeds the probed directories into `spawn_read_skills_from_directories`. Nested provider paths under lazy subdirectories are picked up by the repository watcher when their subtrees load.

**Comment by SagarSDagdu** [comment]
/oz-review Fixed the previously mentioned issue, please re-review

File: app/src/code/file_tree/view.rs [hunk 2]
```diff
@@ -622,9 +622,15 @@ impl FileTreeView {
                 self.rebuild_flattened_items_for_root(repo_path);
                 ctx.notify();
             }
+            RepoMetadataEvent::RepositoryIndexedWithLimit {
+                id: RepositoryIdentifier::Local(_),
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply app/src/code/file_tree/view.rs]: Fixed in 02b24674: gated the toast on whether this view's `root_directories` displays the affected repo (either directly or as the backing entry of an absorbed root), mirroring the `FileTreeEntryUpdated` handler's pattern.

File: app/src/code/file_tree/view.rs [hunk 3]
```diff
@@ -1603,6 +1609,17 @@ impl FileTreeView {
         });
     }
 
+    fn show_repo_indexed_with_limit_toast(ctx: &mut ViewContext<Self>) {
+        let window_id = ctx.window_id();
+        ToastStack::handle(ctx).update(ctx, |toast_stack, ctx| {
+            let toast = DismissibleToast::error(String::from(
+                "Repository is too large to fully index. Subfolders load when expanded.",
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply app/src/code/file_tree/view.rs]: Reproducing requires a >100k-file repo, which I don't have a clean fixture for and is impractical to set up just to grab a screenshot. Repro path: open Warp with a workspace containing >100k tracked files, observe (a) folders now expand instead of being silently empty and (b) a dismissible toast "Repository is too large to fully index. Subfolders load when expanded." — happy to grab a recording if a tester has access to such a repo.

**Comment by SagarSDagdu** [comment]
/oz-review Fixed all the mentioned issues and resolved comments, please re-review

File: app/src/ai/skills/file_watchers/utils.rs [hunk 4]
```diff
@@ -55,6 +55,21 @@ pub fn find_skill_directories_in_tree(
...
 }
 
+/// Returns skill provider directories that exist directly under `repo_path`
+/// (e.g., `<repo>/.agents/skills`, `<repo>/.claude/skills`) via filesystem
+/// probing. Used when the repo metadata tree is shallow (degraded indexing
+/// after `ExceededMaxFileLimit`) and `find_skill_directories_in_tree` cannot
+/// see nested provider paths. Only checks repo-root provider paths; nested
+/// occurrences (`<repo>/sub/.agents/skills`) are picked up later by the
+/// repository file watcher.
+pub fn find_top_level_skill_directories_in_filesystem(repo_path: &Path) -> Vec<PathBuf> {
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply app/src/ai/skills/file_watchers/utils.rs]: Fixed in 1e16b68a: `read_skills_for_repos` now also probes the top-level skill provider paths via the filesystem and dedupes (HashSet) before reading, so cloud-environment skill loading covers degraded repos as well.

File: app/src/code/file_tree/view.rs [hunk 5]
```diff
@@ -1603,6 +1617,17 @@ impl FileTreeView {
...
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply app/src/code/file_tree/view.rs]: Same response as the earlier screenshot ask: reproducing requires a >100k-file repo, which I don't have a clean fixture for and is impractical to spin up just to grab a screenshot. Repro path: open Warp with a workspace containing >100k tracked files (the issue reporter has `~/code` with ~153k); folders that previously appeared empty/non-expandable now expand, and a dismissible toast "Repository is too large to fully index. Subfolders load when expanded." is shown.

[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply app/src/code/file_tree/view.rs]: Okay I have added a loom video in the description, please check.

**Comment by SagarSDagdu** [comment]
/oz-review Fixed all the mentioned issues and resolved comments, also attached a loom video for the demo.
please re-review

File: app/src/ai/skills/file_watchers/utils.rs [hunk 6]
```diff
@@ -55,6 +55,21 @@ pub fn find_skill_directories_in_tree(
...
+    SKILL_PROVIDER_DEFINITIONS
+        .iter()
+        .map(|p| repo_path.join(&p.skills_path))
+        .filter(|p| p.is_dir())
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply app/src/ai/skills/file_watchers/utils.rs]: Good catch. Fixed in 9011b9ae: `find_top_level_skill_directories_in_filesystem` now loads `gitignores_for_directory(repo_path)` and filters candidates with `matches_gitignores(p, true, &gitignores, true)` (`check_ancestors=true`), mirroring `Entry::build_tree`'s filter. Now equivalent to the tree-based path (which uses `get_repo_contents` with `include_ignored=false`).

**Comment by moirahuang** [comment]
Hi @SagarSDagdu, we really appreciate the initiative on the newly filed issue. Per [the contribution guidelines](https://github.com/warpdotdev/warp/blob/master/CONTRIBUTING.md#how-contributing-to-warp-works), the Warp team will manually triage and add a `ready-to-spec` tag when the issue is ready for a spec.

**Comment by SagarSDagdu** [comment]
Hi @moirahuang , thanks for taking a look.
The linked issue with this PR is triaged as a bug, and my assumption was bugs are implicitly ready to implement.
No worries :) what would be the next steps here?

**Comment by moirahuang** [comment]
> Hi @moirahuang , thanks for taking a look.
> The linked issue with this PR is triaged as a bug, and my assumption was bugs are implicitly ready to implement.
> No worries :) what would be the next steps here?

@SagarSDagdu I believe we changed the "bugs are ready to implement" policy as of a few days ago @captainsafia mind confirming please?

**Comment by moirahuang** [comment]
@SagarSDagdu The team has marked the original issue as `ready-to-spec`, mind making a tech spec for the issue please? There's a few things I'd suggest we explore:
1. Should we event attempt to index at all if we know that there are too many files? Or can we just immediately fall back to lazy loading? 
2. Should we show the toast that it's a degraded experience? Or is a toast maybe not necessary?
3. Can we explore the performance implications of lazy loading on a large number of files?

Thank you!

**Comment by SagarSDagdu** [comment]
@moirahuang split the tech spec out into its own PR for separate approval: #10490. This PR (#10234) is the conservative reactive fallback that lands the immediate bug fix; design-direction questions (toast vs persistent indicator, pre-detection, perf telemetry) will be decided on the spec PR and the implementation here will realign once that's approved.

**Comment by SagarSDagdu** [comment]
@moirahuang updated the PR implementation to match what we agreed upon in the spec. Please have a look

**Review by moirahuang — COMMENTED** [review]
overall looks good! mind recording an updated video of the new state and attaching to the PR please?

**Comment by SagarSDagdu** [comment]
@moirahuang Attached a second video in the PR description that shows the updated behavior, and also demonstrated logs of the situation. Review it once.

**Review by moirahuang — APPROVED** [review]
